rust/network/src/reliable_sender.rs

```diff
@@ -0,0 +1,247 @@
+use log::{info, warn};
+use rand::prelude::SliceRandom as _;
+use rand::rngs::SmallRng;
+use rand::SeedableRng as _;
+use std::cmp::min;
+use std::collections::{HashMap, VecDeque};
+use std::fmt::Debug;
+use std::net::SocketAddr;
+use tokio::net::TcpStream;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::sync::oneshot;
+use tokio::time::{sleep, Duration};
+use tokio_util::codec::{Framed, LengthDelimitedCodec};
+
+#[cfg(test)]
+#[path = "tests/reliable_sender_tests.rs"]
+pub mod reliable_sender_tests;
+
+/// Convenient alias for cancel handlers returned to the caller task.
+pub type CancelHandler = oneshot::Receiver<Bytes>;
+
+/// We keep alive one TCP connection per peer, each connection is handled by a separate task (called `Connection`).
+/// We communicate with our 'connections' through a dedicated channel kept by the HashMap called `connections`.
+/// This sender is 'reliable' in the sense that it keeps trying to re-transmit messages for which it didn't
+/// receive an ACK back (until they succeed or are canceled).
+pub struct ReliableSender {
+    /// A map holding the channels to our connections.
+    connections: HashMap<SocketAddr, Sender<InnerMessage>>,
+    /// Small RNG just used to shuffle nodes and randomize connections (not crypto related).
+    rng: SmallRng,
+}
+
+impl std::default::Default for ReliableSender {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ReliableSender {
+    pub fn new() -> Self {
+        Self {
+            connections: HashMap::new(),
+            rng: SmallRng::from_entropy(),
+        }
+    }
+
+    /// Helper function to spawn a new connection.
+    fn spawn_connection(address: SocketAddr) -> Sender<InnerMessage> {
+        let (tx, rx) = channel(1_000);
+        Connection::spawn(address, rx);
+        tx
+    }
+
+    /// Reliably send a message to a specific address.
+    pub async fn send(&mut self, address: SocketAddr, data: Bytes) -> CancelHandler {
+        let (sender, receiver) = oneshot::channel();
+        self.connections
+            .entry(address)
+            .or_insert_with(|| Self::spawn_connection(address))
+            .send(InnerMessage {
+                data,
+                cancel_handler: sender,
+            })
+            .await
+            .expect("Failed to send internal message");
+        receiver
+    }
+
+    /// Broadcast the message to all specified addresses in a reliable manner. It returns a vector of
+    /// cancel handlers ordered as the input `addresses` vector.
+    pub async fn broadcast(
+        &mut self,
+        addresses: Vec<SocketAddr>,
+        data: Bytes,
+    ) -> Vec<CancelHandler> {
+        let mut handlers = Vec::new();
+        for address in addresses {
+            let handler = self.send(address, data.clone()).await;
+            handlers.push(handler);
+        }
+        handlers
+    }
+
+    /// Pick a few addresses at random (specified by `nodes`) and send the message only to them.
```

Comment:
If I call this more than once (upon retry) could I pick the same node twice and unnecessarily try to sync multiple times? Should/Does the caller filter uniques?

Reply:
Yes it is the caller's job